### PR TITLE
Buffs the mind reader mutation to be a bit more consistent/reliable

### PR DESCRIPTION
**Pull request by ATH1909**
## About The Pull Request

The mind reader mutation no longer has a 50% chance to "skip over" a given line when searching for someone's most recent lines of speech. In layman's terms, the mutation will now ALWAYS show you the last three things that your target said (unless they've said less than 3 things so far, in which case it'll just show you those), instead of a semi-random list of the things that your target has said recently.

## Why It's Good For The Game

Now you can determine with 100% certainty that that mime has indeed not spoken so far this round.

It will now also be easier to determine what the subject of your target's last conversation was, allowing you to be much more confident while delivering your smug educated guesses about what someone's been up to recently.

This shouldn't affect the balance of the mind reader mutation too much, as you could effectively get around this RNG-based limitation by just using the power multiple times in a row anyway.

I'm also considering increasing the number of lines revealed by this mutation from 3 lines to 5 lines, to compensate for effectively shortening how far back you can look at your target's conversation history with this mutation.

## Changelog
:cl: ATHATH
balance: The mind reader mutation no longer has a 50% chance to "skip over" a given line when searching for someone's most recent lines of speech. In layman's terms, the mutation will now ALWAYS show you the last three things that your target said (unless they've said less than 3 things so far, in which case it'll just show you those), instead of a semi-random list of the things that your target has said recently.
/:cl:

### CODE REVIEW

**Comment by ATH1909** [comment]
why travis whyyyyyyyyyyy

**Comment by LemonInTheDark** [comment]
There's 2 things here, the fix for say 2 messages in a pool of 5, or 0 in a pool of 1.
Then there's the removal of the chance aspect, which I think adds ~~soul~~ some level of uncertainty, which is nice.

I'd like to have my cake and eat it too, could I have the first and not the second?

**Comment by ATH1909** [comment]
> There's 2 things here, the fix for say 2 messages in a pool of 5, or 0 in a pool of 1.
> Then there's the removal of the chance aspect, which I think adds ~soul~ some level of uncertainty, which is nice.
> 
> I'd like to have my cake and eat it too, could I have the first and not the second?

I added the second one because I thought that there'd be backlash for the first, but I'll undo that, yeah.

**Comment by ATH1909** [comment]
Wait, oh, you're not talking about what I thought you were talking about.

How would I go about implementing the first one but not the second?

**Comment by PKPenguin321** [comment]
this means that if some idiot goes
"HEY"
"YOU"
"STOP"
you will only see those messages ever when mind reading them. having this level of chance gives you the ability to kind of pry further into the past by reading them over and over which was my intent

**Comment by PKPenguin321** [comment]
if you really wanna buff this, lower the instability by like half. i massively overestimated how useful it would be so it's almost never worth it the instability cost

**Comment by ATH1909** [comment]
> this means that if some idiot goes
> "HEY"
> "YOU"
> "STOP"
> you will only see those messages ever when mind reading them. having this level of chance gives you the ability to kind of pry further into the past by reading them over and over which was my intent
> 
> i made this mutation btw

Well, that was why I wanted to increase the maximum number of lines it could tell you to 5, to effectively compensate for that.

When it works, it's a strong mutation, so I think the instability cost is actually justified. Plus, it's one of the only mutations for nonhumans that's actually worth a damn, so if it costed less, I'm pretty sure that you could pick up pretty much every (bar, like, one or two) actually good (so, not geladikinesis or the like) mutation (other than hulk and TK) without going over your instability cap.

**Comment by LemonInTheDark** [comment]
I mean if the length of the printed messages is less then 3, it should add up to that count. That way we can know for sure if that mime spoke this round, and ya don't need to make it just the last 3 lines.

**Comment by ATH1909** [comment]
> I mean if the length of the printed messages is less then 3, it should add up to that count. That way we can know for sure if that mime spoke this round, and ya don't need to make it just the last 3 lines.

But then if they spoke 4 times, there'd be a 0.5^4 chance of no messages being shown...

**Comment by LemonInTheDark** [comment]
Add a check if the total length is more then three, and the returned length is less then three, iterate through the list until you hit a new entry, then insert the new entry into the list.

I'm not a cs major mind, but I'd do something like this

```
build initial list
if(total list.len < 3 || initial list.len >= 3)
    return initial list
var/list/worked_through = list()
var/found_one
for(var/string in total list) //Check every thing they've said
    found_one = TRUE //We're gonna need to know if we found a new one
    for(var/ours in initial list) //Gotta see if we've got one already
        if(ours == string) //If we've got one
            found_one = FALSE //We'll need this in a sec
            break //If we have a copy, no need to go further

    worked through += string //It'll always be behind us when we print this
    if(!found_one) //If it's not a new string
        initial_list -= string //This isn't ahead of us anymore
        continue
    if(worked through.len + initial list.len >= 3) //If we've reached the end, kill it
        break

return worked through + initial list
```

this is strong nightcode, so test it, but I think it'll work. Thoughts?

**Comment by Timberpoes** [comment]
Right, sorry to chime in at the 11th hour but - I don't want to see Mind Reader potentially buffed against antags. There are validhunting uses for Mind Reader as a mutation that make it both strong and desireable to those in the know or with the appropriate disposition towards valids.

I asked a friend to grab more details since Mind Reader's strengths against antags is something I've been aware of but I have never utilised myself.

image

Although it would be a neat usability improvement, I'm not personally looking to expand upon how reliable Mind Reader is. Its unreliability goes at least some way to curbing its strength against certain antags and I don't want to encourage a meta of sec or anyone else all getting Mind Reader because it's now perfectly reliable at all times for catching them valids.

**Comment by ExcessiveUseOfCobblestone** [comment]
It's not too much of a buff considering you can just shift click again.

**Comment by ExcessiveUseOfCobblestone** [comment]
Actually this is a nerf in the sense that the 3 last things is entirely different from previous which would give upto 3 things from EVERYTHING you have ever said (a failed probability would not count against you, it just goes to the next line).

just saying "a" three times is a uncounterable way of btfoing anyone who ever used this mutation looking forward if this was merged.

**Comment by PKPenguin321** [comment]
That's what I've been saying

**Comment by SpaceManiac** [comment]
Four maintainers have read this PR without merging it and the original author of the feature has chimed in against it. It seems there is not interest in merging this change.
